### PR TITLE
#6521 History client middleware

### LINKED ISSUE #6521: History Client Middleware
Browsers Keep a history of calls made by the browser. Would be nice if we had a client middleware which did similar in case we ever want to show and capture the interactions that a specific flow may have generated without leaking confidential information like headers/cookies/etc.

```scala
// Browser save Title HTML and Favicon - Maybe a later pass maybe just options that are none for now to keep bincompat
// Maybe we want a builder class for this so we can add additional information to it.

def middleware[F[_]: MonadCancelThrow](client: Client[F], history: Ref[F, List[(HttpDate, Method, Uri)], maxSize: Int): Client[F] = Client[F]{(req: Request[F]) =>
  Resource.eval(req.headers.get[Date].fold(HttpDate.now[F])(d => d.date.pure[F]).flatMap{ date => 
     val allTheOtherStuff 
     history.modify(l => event :: l) >>
     dropElementsAfterN >>
     clinet.run(req)
  }
} // obviously psuedo
```

### PR DESCRIPTION
**Pull request by SallyPerez**
Fixes #6521

### CODE REVIEW

File: client/shared/src/main/scala/org/http4s/client/middleware/History.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.client.middleware
+
+import cats.effect._
+import cats.implicits.catsSyntaxApplicativeId
```
Comment (danicheg) [inline client/shared/src/main/scala/org/http4s/client/middleware/History.scala:20]: I think here better to use the following import
```suggestion
import cats.syntax.applicative._
```

File: client/shared/src/main/scala/org/http4s/client/middleware/History.scala [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import org.http4s._
+import org.http4s.client.Client
+import org.http4s.headers.Date
+
+/** This Middleware provides history tracking of Uri's visited.
+  * History entries are ordered most recent to oldest.
+  *
+  * @param client: Client[F]
+  * @param history: Ref[F, Vector[HistoryEntry]
+  * @param maxSize: Int
```
Comment (danicheg) [inline client/shared/src/main/scala/org/http4s/client/middleware/History.scala:30]: It seems like these parameters are missing in the constructor

File: client/shared/src/main/scala/org/http4s/client/middleware/History.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  */
+
+class HistoryEntry private (val httpDate: HttpDate, val method: Method, val uri: Uri) {
```
Comment (danicheg) [inline client/shared/src/main/scala/org/http4s/client/middleware/History.scala:33]: A controversial opinion maybe, but I like to mark every class `final` whenever it makes sense

File: client/shared/src/main/scala/org/http4s/client/middleware/History.scala [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+object HistoryEntry {
+  def apply(date: HttpDate, method: Method, uri: Uri): HistoryEntry =
+    new HistoryEntry(date, method, uri)
+
+}
+
+final class HistoryBuilder[F[_]: MonadCancelThrow: Clock] private (
+    private val client: Client[F],
+    private val history: Ref[F, Vector[HistoryEntry]],
```
Comment (danicheg) [inline client/shared/src/main/scala/org/http4s/client/middleware/History.scala:63]: The whole constructor is `private`, so maybe we can omit these `private` ascriptions

File: client/shared/src/test/scala/org/http4s/client/middleware/HistorySuite.scala [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      override def realTime: IO[FiniteDuration] =
+        IO.pure(FiniteDuration(0L, scala.concurrent.duration.HOURS))
+    }
+
+    val expected: Vector[HistoryEntry] = Vector(
+      HistoryEntry(req3.headers.get[Date].get.date, req3.method, req3.uri),
+      HistoryEntry(req2.headers.get[Date].get.date, req2.method, req2.uri),
+    )
+
+    implicitly[Clock[IO]]
```
Comment (danicheg) [inline client/shared/src/test/scala/org/http4s/client/middleware/HistorySuite.scala:127]: Maybe this line is redundant

**Review by danicheg — COMMENTED** [review]
My few comments

File: client/shared/src/main/scala/org/http4s/client/middleware/History.scala (hunk 3 above)
Comment (rossabaker) [reply client/shared/src/main/scala/org/http4s/client/middleware/History.scala:33]: The reason this isn't a case class is to reserve space for bincompat evolution.  Would we get more flexibility making it a sealed trait with a single private implementation?   I can't recall all the ramifications of adding a private impl later.

But if not sealed, yes, private.

File: client/shared/src/main/scala/org/http4s/client/middleware/History.scala (hunk 4 above)
Comment (rossabaker) [reply client/shared/src/main/scala/org/http4s/client/middleware/History.scala:63]: That would make `client` and `history` publicly accessible after construction, like `maxSize`.  The question is should the properties be readable before the builder is built?  I'm ambivalent, but I don't see any reason for them not to be the same.

File: client/shared/src/main/scala/org/http4s/client/middleware/History.scala (hunk 2 above)
Comment (hamnis) [reply client/shared/src/main/scala/org/http4s/client/middleware/History.scala:30]: should be moved to the correct place

File: client/shared/src/main/scala/org/http4s/client/middleware/History.scala (hunk 1 above)
Comment (SallyPerez) [reply client/shared/src/main/scala/org/http4s/client/middleware/History.scala]: updated, thanks!

File: client/shared/src/test/scala/org/http4s/client/middleware/HistorySuite.scala (hunk 5 above)
Comment (SallyPerez) [reply client/shared/src/test/scala/org/http4s/client/middleware/HistorySuite.scala]: removed, thank you

File: client/shared/src/main/scala/org/http4s/client/middleware/History.scala (hunk 3 above)
Comment (SallyPerez) [reply client/shared/src/main/scala/org/http4s/client/middleware/History.scala]: made HistoryEntry final, thanks

**Review by rossabaker — APPROVED** [review]

**Comment by danicheg** [comment]
I flipped through the changes, and with Ross's approval, it's good to go — no blockers from me

**Pull request merged**